src/core/bitset.c

```diff
@@ -0,0 +1,649 @@
+    const igraph_integer_t final_block_size = bitset->size % IGRAPH_INTEGER_SIZE ? bitset->size % IGRAPH_INTEGER_SIZE : IGRAPH_INTEGER_SIZE;
+    const igraph_integer_t slots = IGRAPH_BIT_NSLOTS(bitset->size);
+    const igraph_integer_t one = 1, zero = 0; /* to avoid the need to cast 1 and 0 to igraph_integer_t below */
+    const igraph_integer_t mask = final_block_size == IGRAPH_INTEGER_SIZE ? ~zero : ((one << final_block_size) - one);
+
+    for (igraph_integer_t i = 0; i + 1 < slots; ++i) {
+        if (VECTOR(*bitset)[i] != zero) {
+            return IGRAPH_INTEGER_SIZE * i + IGRAPH_CTZ(VECTOR(*bitset)[i]);
+        }
+    }
+    if (bitset->size && (mask & VECTOR(*bitset)[slots - 1]) != zero) {
+        return IGRAPH_INTEGER_SIZE * (slots - 1) + IGRAPH_CTZ(mask & VECTOR(*bitset)[slots - 1]);
+    }
+
+    return bitset->size;
+}
+
+/**
+ * \ingroup bitset
+ * \function igraph_bitset_countr_one
+ * \brief The number of trailing ones in the bitset.
+ *
+ * \experimental
+ *
+ * Returns the number of trailing ones (starting at the least significant bit)
+ * in the bitset before the first zero is encountered.
+ * If the bitset is all ones, then its size is returned.
+ *
+ * \param bitset The bitset object
+ * \return The number of trailing ones in the bitset.
+ *
+ * Time complexity: O(n/w).
+ */
+
+igraph_integer_t igraph_bitset_countr_one(const igraph_bitset_t *bitset) {
+    const igraph_integer_t final_block_size = bitset->size % IGRAPH_INTEGER_SIZE ? bitset->size % IGRAPH_INTEGER_SIZE : IGRAPH_INTEGER_SIZE;
+    const igraph_integer_t slots = IGRAPH_BIT_NSLOTS(bitset->size);
+    const igraph_integer_t one = 1, zero = 0; /* to avoid the need to cast 1 and 0 to igraph_integer_t below */
+    const igraph_integer_t mask = final_block_size == IGRAPH_INTEGER_SIZE ? zero : ~((one << final_block_size) - one);
+
+    for (igraph_integer_t i = 0; i + 1 < slots; ++i) {
+        if (VECTOR(*bitset)[i] != ~zero) {
+            return IGRAPH_INTEGER_SIZE * i + IGRAPH_CTO(VECTOR(*bitset)[i]);
+        }
+    }
+    if (bitset->size && (mask | VECTOR(*bitset)[slots - 1]) != ~zero) {
+        return IGRAPH_INTEGER_SIZE * (slots - 1) + IGRAPH_CTO(mask | VECTOR(*bitset)[slots - 1]);
+    }
+
+    return bitset->size;
+}
+
+/**
+ * \ingroup bitset
+ * \function igraph_bitset_or
+ * \brief Bitwise OR of two bitsets.
+ *
+ * \experimental
+ *
+ * Applies a bitwise or to the contents of two bitsets and stores it in an
+ * already initialized bitset. The destination bitset may be equal to one
+ * (or even both) of the sources.
+ *
+ * \param dest The bitset object where the result is stored
+ * \param src1 A bitset
+ * \param src2 A bitset
+ *
+ * Time complexity: O(n/w).
+ */
+
+void igraph_bitset_or(igraph_bitset_t *dest,
+                      const igraph_bitset_t *src1, const igraph_bitset_t *src2) {
+    for (igraph_integer_t i = 0; i < IGRAPH_BIT_NSLOTS(dest->size); ++i) {
+        VECTOR(*dest)[i] = VECTOR(*src1)[i] | VECTOR(*src2)[i];
+    }
+}
```

Comment:
This and other similar functions should check that the two sources have compatible sizes.